README.md
<!--
 * @Author: your name
 * @Date: 2020-03-08 19:29:37
 * @LastEditTime: 2020-03-15 09:44:05
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit 
 * @FilePath: \DevUI-Language-Support\README.md
 -->
# 一个基于LSP的DevUI语言组件

![demo](https://s2.ax1x.com/2020/03/08/3z184H.gif)

#### 说明

这个组件面向DevUI组件库的开发者，将提供有关于DevUI开源组件库的一系列帮助，包括:
- 代码提示
- 代码纠错
- 悬浮提醒
- 跳转帮助
(以上内容均在开发中)

#### 下载方式

- 您可以在<a herf="https://marketplace.visualstudio.com/items?itemName=yqLiu.devui-language-support&ssr=false">vscode的插件库</a>进行下载。

- 若要下载不同的版本，请访问<a herf="https://github.com/sspku-yqLiu/DevUI-language-support/releases/">Release</a>寻求进一步的帮助。

#### 帮助我们进行改进

- 非常欢迎您提交bug及改良的建议，您可以<a herf="https://github.com/sspku-yqLiu/DevUI-language-support/issues">点此提交</a>。

- 您也可以直接进入到<a herf="https://github.com/sspku-yqLiu/DevUI-language-support">我们的仓库</a>，提交一份属于您的代码！

#### 关于DevUI

- DevUI官方网址:https://devui.design/home
- DevUI组件库:https://github.com/sspku-yqLiu/ng-devui

CHANGELOG.md
<!--
 * @Author: your name
 * @Date: 2020-03-14 17:21:08
 * @LastEditTime: 2020-03-15 19:11:34
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \DevUI-Language-Support\CHANGELOG.md
 -->
# 2020/3/15 V 0.2.1~V0.2.2

### 完成点：
V0.2.1
- 添加了打包脚本，现在，你可以使用'yarn package'命令对项目进行一键打包。
- 更新了README.md

v0.2.2
- 优化了打包脚本，现在，插件的体积将是原来的60%左右

### 架构更新

- 将采用typescript所提供的Language-servicePlugin，为devui提供包括自动提示，代码纠错，悬浮提醒的一系列功能。点击<a herf="https://github.com/sspku-yqLiu/devui-language-service">这里</a>了解更多

- 在完成service后,他将集成到插件的devui库中，注意：这并不意味着要求用户的devui中也同样需要安装language-service服务，但是我强烈推荐devui集成语言服务，因为这将会大大帮助插件开发者。

BUGS :
- 参考上一版，因为是周末就没怎么敲代码了=-=

# 2020/3/14 V0.2.0

### 完成点：
- 使用了angular-language-support架构
- 添加了拦截器功能，现在在没有DevUI依赖的文件夹中，插件将不会被激活。
- 使用rollup对项目进行打包，减少了代码冗余。
- 添加了Logo License 以CHANGELOG.md

### BUGS:
- 使用人工打包而非批处理打包，java文件无法满足自动打包功能。
- 似乎disableLanguageService()函数并没有起作用，也许只在DEBUG情境下才这样
- 仍然没有添加对typescript语法的支持
- package.json文件中的main在DEBUG模式与生产模式中不同，这是否是打包批处理文件应该完成的工作？

# 2020/3/12 V0.1.0

完成点：

- 完成最小可用产品，能够在html中实现大类标签提示。

BUGS：

- 在所有的html中都会提示。冗余严重

下一步要做的事：

- 完成大类中api的提示
- 完成启动控制器，仅当发现ng-devui插件时启动提示功能。

# 2020/3/11 V0.0.1

完成点：

正式登陆vscode 完成在商店中发布

BUGS：

无法运行，初始化失败，怀疑是session类的BUG

下一步要做的事：

发布最小可用产品。

client/src/extension.ts
/*
 * @Author: your name
 * @Date: 2020-03-04 17:55:03
 * @LastEditTime: 2020-03-14 11:34:21
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \lsp-sample\client\src\extension.ts
 */
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import * as lsp from 'vscode-languageclient';

import { registerCommands } from './commands';
import { projectLoadingNotification } from './protocol';

// import {
// 	LanguageClient,
// 	LanguageClientOptions,
// 	ServerOptions,
// 	TransportKind,
// 	RevealOutputChannelOn
// } from 'vscode-languageclient';

// let client: LanguageClient;

export function activate(context: vscode.ExtensionContext) {
	// The server is implemented in node
	// let serverModule = context.asAbsolutePath(
	// 	path.join('server', 'out', 'server.js')
	// );
	// The debug options for the server
	// --inspect=6009: runs the server in Node's Inspector mode so VS Code can attach to the server for debugging
	// let debugOptions = { execArgv: ['--nolazy', '--inspect=6009'] };

	// If the extension is launched in debug mode then the debug server options are used
	// Otherwise the run options are used
	const serverOptions: lsp.ServerOptions = {
		run: getServerOptions(context, false /* debug */),
		debug: getServerOptions(context, true /* debug */),
	};

	// Options to control the language client
	const clientOptions: lsp.LanguageClientOptions = {
		// Register the server for plain text documents
		documentSelector: [
			{ scheme: 'file', language: 'html' },
			{ scheme: 'file', language: 'typescript' },
		],
		synchronize: {
			// Notify the server about file changes to '.clientrc files contained in the workspace
			// fileEvents: vscode.workspace.createFileSystemWatcher('**/.clientrc')
			/* 为什么这个地方要用tsconfig? */
			fileEvents: vscode.workspace.createFileSystemWatcher('**/tsconfig.json')
		},
		revealOutputChannelOn: lsp.RevealOutputChannelOn.Never
	};

	// Create the language client and start the client.
	const forceDebug = process.env['NG_DEBUG'] === 'true';
	/* 名字和id 不影响提示效果*/
	const client = new lsp.LanguageClient(
		// 'languageServerExample',
		// 'Language Server Example',
		'devui-language-support-client',
		// 'DevUI_Language_Support',
		serverOptions,
		clientOptions,
		forceDebug
	);

	/* 保证客户端能够在扩展关闭的同时关闭，并开启客户端*/
	context.subscriptions.push(...registerCommands(client), client.start());

	/**
	 * @description: 对client的变化进行应答 
	 * @param {type} 
	 * @return: 
	 */
	client.onDidChangeState((e) => {
		let task: { resolve: () => void } | undefined;
		if (e.newState == lsp.State.Running) {
			client.onNotification(projectLoadingNotification.start, () => {
				if (task) {
					task.resolve();
					task = undefined;
				}
				vscode.window.withProgress({
					location: vscode.ProgressLocation.Window,
					title: 'Initialzing DevUI language features',
				},
					() => new Promise((resolve) => {
						task = { resolve };
					}));
			});
			client.onNotification(projectLoadingNotification.finish, () => {
				if (task) {
					task.resolve();
					task = undefined;
				}
			});
		}
	});
}
/**
 * Return the paths for the module that corresponds to the specified `configValue`,
 * and use the specified `bundled` as fallback if none is provided.
 * @param configName
 * @param bundled
 */
function getProbeLocations(configValue: string | null, bundled: string): string[] {
	const locations = [];
	// Always use config value if it's specified
	if (configValue) {
		locations.push(configValue);
	}
	// Prioritize the bundled version
	locations.push(bundled);
	// Look in workspaces currently open
	const workspaceFolders = vscode.workspace.workspaceFolders || [];
	for (const folder of workspaceFolders) {
		locations.push(folder.uri.fsPath);
	}
	return locations;
}

/** 
 * Construct the arguments that's used to spawn the server process.
 * @param ctx vscode extension context
 * @param debug true if debug mode is on
 */
function constructArgs(ctx: vscode.ExtensionContext, debug: boolean): string[] {
	const config = vscode.workspace.getConfiguration();
	const args: string[] = [];
	const ngLog: string = config.get('angular.log', 'off');
	if (ngLog !== 'off') {
		// Log file does not yet exist on disk. It is up to the server to create the file.
		const logFile = path.join(ctx.logPath, 'nglangsvc.log');
		args.push('--logFile', logFile);
		args.push('--logVerbosity', debug ? 'verbose' : ngLog);
	}

	const ngdk: string | null = config.get('angular.ngdk', null);
	const ngProbeLocations = getProbeLocations(ngdk, ctx.asAbsolutePath('server'));
	args.push('--ngProbeLocations', ngProbeLocations.join(','));

	const tsdk: string | null = config.get('typescript.tsdk', null);
	const tsProbeLocations = getProbeLocations(tsdk, ctx.extensionPath);
	args.push('--tsProbeLocations', tsProbeLocations.join(','));

	/*devui探针*/
	// const devuidk: string|null = config.get('ng-devui',null);
	// const devuiProbeLocations = getProbeLocations(devuidk,ctx.extensionPath);
	// args.push('--devuiProbeLocations',devuiProbeLocations.join(','));
	return args;
}

function getServerOptions(ctx: vscode.ExtensionContext, debug: boolean): lsp.NodeModule {
	// Environment variables for server process
	const prodEnv = {
		// Force TypeScript to use the non-polling version of the file watchers.
		TSC_NONPOLLING_WATCHER: true,
	};
	const devEnv = {
		...prodEnv,
		NG_DEBUG: true,
	};

	// Node module for the language server
	const prodBundle = ctx.asAbsolutePath('server');
	const devBundle = ctx.asAbsolutePath(path.join('server', 'out', 'server.js'));

	// Argv options for Node.js
	const prodExecArgv: string[] = [];
	const devExecArgv: string[] = [
		// do not lazily evaluate the code so all breakpoints are respected
		'--nolazy',
		// If debugging port is changed, update .vscode/launch.json as well
		'--inspect=6009',
	];

	return {
		// VS Code Insider launches extensions in debug mode by default but users
		// install prod bundle so we have to check whether dev bundle exists.
		module: debug && fs.existsSync(devBundle) ? devBundle : prodBundle,
		transport: lsp.TransportKind.ipc,
		args: constructArgs(ctx, debug),
		options: {
			env: debug ? devEnv : prodEnv,
			execArgv: debug ? devExecArgv : prodExecArgv,
		},
	};
}
// export function deactivate(): Thenable<void> | undefined {
// 	if (!client) {
// 		return undefined;
// 	}
// 	return client.stop();
// }
